src/main/java/org/opensearch/geospatial/ip2geo/processor/Ip2GeoProcessor.java

```diff
@@ -206,14 +169,45 @@ public void onFailure(final Exception e) {
         });
     }
 
+    @VisibleForTesting
+    protected void handleSingleIp(
+        final String ip,
+        final Map<String, Object> ipToGeoData,
+        final IngestDocument ingestDocument,
+        final BiConsumer<IngestDocument, Exception> handler
+    ) {
+        if (!ipToGeoData.isEmpty()) {
+            ingestDocument.setFieldValue(targetField, filteredGeoData(ipToGeoData, ip));
+        }
+        handler.accept(ingestDocument, null);
+    }
+
+    private Map<String, Object> filteredGeoData(final Map<String, Object> ipToGeoData, final String ip) {
+        Map<String, Object> filteredGeoData;
+        if (properties == null) {
+            return ipToGeoData;
+        }
+
+        filteredGeoData = new HashMap<>();
+        for (String property : this.properties) {
+            if (property.equals(PROPERTY_IP)) {
+                filteredGeoData.put(PROPERTY_IP, ip);
+            } else {
+                filteredGeoData.put(property, ipToGeoData.get(property));
+            }
```

Comment:
This can be simplified. 
1. filteredGeoData = properties.stream().filter("not ip").collect(Collectors.toMap(...)
2. if properties contains "ip", filteredGeoData.put(PROPERTY_IP, ip);

Reply:
I am not convinced that suggested code is simpler than the original code..

Reply:
What you were doing is imperative style. The suggested code is declarative. Declarative style conveys your logic better than existing way of mixing both logic in conditional statements.

Reply:
Fair enough.